alchemiscale/compute/service.py

```diff
@@ -74,59 +78,143 @@ def __init__(
 
         if scopes is None:
             self.scopes = [Scope()]
+        else:
+            self.scopes = scopes
+
+        self.shared_basedir = Path(shared_basedir)
+        self.shared_basedir.mkdir(exist_ok=True)
+        self.keep_shared = keep_shared
+
+        self.scratch_basedir = Path(scratch_basedir)
+        self.scratch_basedir.mkdir(exist_ok=True)
+        self.keep_scratch = keep_scratch
 
-        self.shared = shared_path
         self.scheduler = sched.scheduler(time.monotonic, time.sleep)
 
+        self.compute_service_id = ComputeServiceID(f"{self.name}-{uuid4()}")
+
+        self.int_sleep = InterruptableSleep()
+        self.logger = logging.getLogger("AlchemiscaleSynchronousComputeService")
+        self.logger.setLevel(loglevel)
+
+        self.logger.addHandler(logging.StreamHandler())
+
         self._stop = False
 
-    def heartbeat(self):
+    def _register(self):
+        """Register this compute service with the compute API."""
+        self.client.register(self.compute_service_id)
+
+    def _deregister(self):
+        """Deregister this compute service with the compute API."""
+        self.client.deregister(self.compute_service_id)
+
+    def beat(self):
         """Deliver a heartbeat to the compute API, indicating this service is still alive."""
-        ...
+        self.client.heartbeat(self.compute_service_id)
+        self.logger.info("Updated heartbeat")
+
+    def heartbeat(self):
+        """Start up the heartbeat, sleeping for `self.heartbeat_interval`"""
+        while True:
+            if self._stop:
+                break
+            self.beat()
+            time.sleep(self.heartbeat_interval)
 
-    def get_tasks(self, count=1) -> List[Optional[ScopedKey]]:
+    def claim_tasks(self, count=1) -> List[Optional[ScopedKey]]:
         """Get a Task to execute from compute API.
 
         Returns `None` if no Task was available matching service configuration.
 
         """
+        # list of tasks to return
+        tasks = []
+
         taskhubs: Dict[ScopedKey, TaskHub] = self.client.query_taskhubs(
             scopes=self.scopes, return_gufe=True
         )
 
-        # based on weights, choose taskhub to draw from
-        taskhub: List[ScopedKey] = random.choices(
-            list(taskhubs.keys()), weights=[tq.weight for tq in taskhubs.values()]
-        )[0]
+        if len(taskhubs) == 0:
+            return []
 
-        # claim tasks from the taskhub
-        tasks = self.client.claim_taskhub_tasks(
-            taskhub, claimant=self.name, count=count
-        )
+        # claim tasks from taskhubs based on weight; keep going till we hit our
+        # total desired task count, or we run out of taskhubs to draw from
+        while len(tasks) < count and len(taskhubs) > 0:
+            # based on weights, choose taskhub to draw from
+            taskhub: List[ScopedKey] = random.choices(
+                list(taskhubs.keys()), weights=[tq.weight for tq in taskhubs.values()]
```

Comment:
nitpick: th clearer than tq
```suggestion
                list(taskhubs.keys()), weights=[th.weight for th in taskhubs.values()]
```

Reply:
Ah, thank you! Yes this is a leftover from the old `TaskQueue` implementation.